### PR TITLE
Issue #2281111: Porting 'Entity is of bundle (entity_is_of_bundle)' cond...

### PR DESCRIPTION
**Pull request by ndewhurst**
...ition to 8.x

See: d.o issue #2281111 - https://drupal.org/node/2281111
This creates an EntityIsOfBundle condition and an associated test that checks a simulated node entity's type and bundle.

### CODE REVIEW

File: src/Plugin/Condition/EntityIsOfBundle.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin ID for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\TypedData\TypedDataManager $typed_data_manager
+   *   The typed data manager.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, TypedDataManager $typed_data_manager) {
```
Comment (fago) [inline src/Plugin/Condition/EntityIsOfBundle.php]: Does nothing different than the parent, thus needs to go away.

File: src/Plugin/Condition/EntityIsOfBundle.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function contextDefinitions(TypedDataManager $typed_data_manager) {
+    $contexts['entity'] = ContextDefinition::create($typed_data_manager, 'entity')
+      ->setLabel(t('Entity'))
+      ->setDescription(t('Specifies the entity for which to evaluate the condition.'));
+
+    // @todo: specify input mechanisms and/or constraints once configuration/UI
```
Comment (fago) [inline src/Plugin/Condition/EntityIsOfBundle.php]: Constraints we can specify already, so it should be done now. Also, let's just add todo line for each todo item + start each sentence uppercase end end it with a trailing point.

File: src/Plugin/Condition/EntityIsOfBundle.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    //  questions are settled.
+    $contexts['type'] = ContextDefinition::create($typed_data_manager, 'string')
+      ->setLabel(t('Type'))
+      ->setDescription(t('The type of the evaluated entity.'));
+
+    // @todo: specify input mechanisms and/or constraints once configuration/UI
```
Comment (fago) [inline src/Plugin/Condition/EntityIsOfBundle.php]: same as above

File: src/Plugin/Condition/EntityIsOfBundle.php [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    //  questions are settled.
+    $contexts['bundle'] = ContextDefinition::create($typed_data_manager, 'string')
+      ->setLabel(t('Bundle'))
+      ->setDescription(t('The bundle of the evaluated entity.'));
+
+    return $contexts;
+  }
+
```
Comment (fago) [inline src/Plugin/Condition/EntityIsOfBundle.php:62]: one empty line too many

File: src/Plugin/Condition/EntityIsOfBundle.php [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+   */
+  public function summary() {
+    return $this->t('Entity is of bundle');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function evaluate() {
+    // Load the entity to evaluate.
```
Comment (fago) [inline src/Plugin/Condition/EntityIsOfBundle.php]: unnecessary comment, reading the code tells me that.

File: src/Plugin/Condition/EntityIsOfBundle.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    $providedEntity = $this->getContextValue('entity');
+    // Retrieve the entity type specified as part of this condition.
```
Comment (fago) [inline src/Plugin/Condition/EntityIsOfBundle.php]: unnecessary comment, reading the code tells me that.

File: src/Plugin/Condition/EntityIsOfBundle.php [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    $specified_type = $this->getContextValue('type');
+    // Retrieve the entity bundle specified as part of this condition.
```
Comment (fago) [inline src/Plugin/Condition/EntityIsOfBundle.php]: unnecessary comment, reading the code tells me that.

File: src/Plugin/Condition/EntityIsOfBundle.php [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    $specified_bundle = $this->getContextValue('bundle');
+    // Retrieve the type and bundle of the evaluated entity.
```
Comment (fago) [inline src/Plugin/Condition/EntityIsOfBundle.php]: unnecessary comment, reading the code tells me that.

File: src/Plugin/Condition/EntityIsOfBundle.php [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    $entity_type = $providedEntity->getEntityTypeId();
+    $entity_bundle = $providedEntity->bundle();
+    // Check to see whether the entity's bundle and type match the specified
+    //  values.
```
Comment (fago) [inline src/Plugin/Condition/EntityIsOfBundle.php:80]: one white space too many

File: src/Plugin/Condition/EntityIsOfBundle.php (hunk 2 above)
Comment (ndewhurst) [inline src/Plugin/Condition/EntityIsOfBundle.php]: OK, cool. I'm not familiar with constraints, but the concept makes sense. It looks like this is the authoritative reference for available constraints: https://api.drupal.org/api/drupal/namespace/Symfony%21Component%21Validator%21Constraints/8
What constraints would be relevant here? "NotBlank"...or "NotNull" may be relevant for all three. "Type" seems redundant. "Length" would make sense for the strings...

**Comment by fubhy** [comment]
Thanks, merged! I swapped the simple test out for a unit test while merging.

**Pull request closed without merging**